Maths/FindMedian.java

```diff
@@ -0,0 +1,133 @@
+package Maths;
+import java.util.Arrays;
+
+/**
+ * Implementation of a median algorithm.
+ * @author github.com/b3z
+ */
+public class FindMedian {
+
+    /**
+     * Computes and retrieves the lower median of the given array of pairwise
+     * distinct numbers.
+     * 
+     * @param numbers array with pairwise distinct numbers.
+     * @return the lower median.
+     * @throw IllegalArgumentException if the array is {@code null} or empty.
+     */
+    public static int lowerMedian(int[] numbers) {
+
+        if (numbers == null || numbers.length == 0) // validate input.
+            throw new IllegalArgumentException();
+
+        return findMedian(numbers, ((numbers.length + 1) / 2) - 1); // compute median of array.
+    }
+
+    /**
+     * Partitions 5er groups and recalls.
+     * 
+     * @param numbers     input array.
+     * @param indexMedian index of median to search by.
+     * @return sorted by pivot. Also called median.
+     */
+    public static int findMedian(int[] numbers, int indexMedian) {
+
+        int pivot;
+        if (numbers.length <= 5)
+            pivot = subMedian(numbers);
+        else {
+            int groups = numbers.length / 5; // Define how many groups are necessary.
+            groups += (numbers.length % 5 != 0) ? 1 : 0; // Handle incomplete groups (simply add one).
+
+            int[] submedians = grouping(groups, numbers); // make 5er groups and save submedian.
+
+            pivot = (submedians.length <= 5) ? subMedian(submedians) // if only 5 are left calculate median.
+                    : findMedian(submedians, ((submedians.length - 1) / 2)); // else repeat.
+        }
+        return splitLists(numbers, pivot, indexMedian);
+    }
+
+    /**
+     * Groups the array and saves their submedians.
+     * 
+     * @param size    groups we have.
+     * @param numbers original numbers to group.
+     * @return submedians
+     */
+    private static int[] grouping(int size, int[] numbers) {
+        int[] submedians = new int[size];
+        int[] tmp = new int[5];
+        int pointer = 0;
+
+        for (int i = 0; i < numbers.length / 5; i++) {
+            for (int k = 0; k < 5; k++)
+                tmp[k] = numbers[pointer++];
+
+            submedians[i] = subMedian(tmp);
+        }
+
+        if (numbers.length % 5 != 0) {  // handle groups smaller than 5.
+            int[] subGroup = new int[numbers.length % 5];
+            for (int i = 0; i < (numbers.length % 5); i++)
+                subGroup[i] = numbers[pointer++];
```

Comment:
seperate statements `tmp[k] = numbers[pointer++];`
i.e.
`pointer++;`
`numbers[pointer];`

Reply:
Yes, would make it more readable.

Reply:
> i.e.
> `pointer++;`
> `numbers[pointer];`

This example is wrong. Its post increment so you can't just execute it before reading the value.

In the code it is meant to be 

```Java
numbers[pointer++];
```

This would be 

```Java
numbers[pointer];
pointer++;

Reply:
With these changes nobody else will misinterprete it like I did :)